src/transform/src/attribute/cardinality.rs

```diff
@@ -0,0 +1,646 @@
+            | TableFunc::GenerateSeriesTimestamp
+            | TableFunc::GenerateSeriesTimestampTz => input.clone(),
+            TableFunc::Wrap { types, width } => {
+                input * (f64::cast_lossy(types.len()) / f64::cast_lossy(*width))
+            }
+            _ => {
+                // TODO(mgree) what explosion factor should we make up?
+                input * &SymExp::from(4.0)
+            }
+        }
+    }
+
+    fn predicate(&self, expr: &MirScalarExpr, indexed_columns: &BTreeSet<usize>) -> SymExp {
+        let index_cardinality = |expr: &MirScalarExpr| -> Option<SymExp> {
+            match expr {
+                MirScalarExpr::Column(col) => {
+                    if indexed_columns.contains(col) {
+                        Some(SymbolicExpression::symbolic(FactorizerVariable::Index(
+                            *col,
+                        )))
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            }
+        };
+
+        match expr {
+            MirScalarExpr::Column(_)
+            | MirScalarExpr::Literal(_, _)
+            | MirScalarExpr::CallUnmaterializable(_) => SymExp::from(1.0),
+            MirScalarExpr::CallUnary { func, expr } => match func {
+                UnaryFunc::Not(_) => 1.0 - self.predicate(expr, indexed_columns),
+                UnaryFunc::IsTrue(_) | UnaryFunc::IsFalse(_) => SymExp::from(0.5),
+                UnaryFunc::IsNull(_) => {
+                    if let Some(icard) = index_cardinality(expr) {
+                        icard
+                    } else {
+                        SymExp::from(WORST_CASE_SELECTIVITY)
+                    }
+                }
+                _ => SymExp::from(WORST_CASE_SELECTIVITY),
+            },
+            MirScalarExpr::CallBinary { func, expr1, expr2 } => {
+                match func {
+                    BinaryFunc::Eq => match (index_cardinality(expr1), index_cardinality(expr2)) {
+                        (Some(icard1), Some(icard2)) => SymbolicExpression::max(icard1, icard2),
+                        (Some(icard), None) | (None, Some(icard)) => icard,
+                        (None, None) => SymExp::from(WORST_CASE_SELECTIVITY),
+                    },
+                    // 1.0 - the Eq case
+                    BinaryFunc::NotEq => match (index_cardinality(expr1), index_cardinality(expr2))
+                    {
+                        (Some(icard1), Some(icard2)) => {
+                            1.0 - SymbolicExpression::max(icard1, icard2)
+                        }
+                        (Some(icard), None) | (None, Some(icard)) => 1.0 - icard,
+                        (None, None) => SymExp::from(1.0 - WORST_CASE_SELECTIVITY),
+                    },
+                    BinaryFunc::Lt | BinaryFunc::Lte | BinaryFunc::Gt | BinaryFunc::Gte => {
+                        // if we have high/low key values and one of the columns is an index
```

Comment:
How do you assert the condition from this comment in the code here? AFACT this will return 0.33 for every binary expression for one of the above four functions.

Reply:
Comment was incomplete 😓 . Should have read:

```rust
BinaryFunc::Lt | BinaryFunc::Lte | BinaryFunc::Gt | BinaryFunc::Gte => {
  // TODO(mgree) if we have high/low key values and one of the columns is an index, we can do better
  SymExp::from(0.33)
}
```

With those stats, you can make a guess about distribution.